src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs

```diff
@@ -0,0 +1,87 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace Microsoft.DotNet.Build.CloudTestTasks
+{
+    public sealed class FilterForUpload : Task
+    {
+        /// <summary>
+        ///     This is the list of test archives produced with OSPlatformConfig metadata set
+        /// </summary>
+        [Required]
+        public ITaskItem[] UnfilteredUploadItems { get; set; }
+
+        /// <summary>
+        ///     This is the TestTFM folder that we want to select test zips from. This needs to be on the path to the test zip.
+        /// </summary>
+        [Required]
+        public string FilterToTestTFM { get; set; }
+
+        /// <summary>
+        ///     This is the OSGroup folder that we want to select test zips from. This does not require an exact match as some
+        ///     tests
+        ///     may or may not have an exact OSGroup folder match.
+        /// </summary>
+        [Required]
+        public string FilterToOSGroup { get; set; }
+
+        /// <summary>
+        ///     This is the bin folder that should be removed from all paths.
+        /// </summary>
+        [Required]
+        public string BinDir { get; set; }
+
+        [Output]
+        public ITaskItem[] FilteredUploadItems { get; set; }
+
+        public override bool Execute()
+        {
+            List<ITaskItem> filteredBasedOnTestTfm =
+                UnfilteredUploadItems.Where(
+                    item => item.ItemSpec.Replace(BinDir, string.Empty).Contains(FilterToTestTFM)).ToList();
```

Comment:
Use this instead of "thing.Replace"

``` csharp
string myRandomPath = ...;
string binDir = ...;

binDir = Path.GetFullPath(binDir);
string fullRandomPath = Path.GetFullPath(myRandomPath);
if (fullRandomPath.StartsWith(binDir)) {
  return fullRandomPath.Substring(binDir.Length + 1)
else
  return myRandomPath;
```